### PR TITLE
bpo-33178: Add BigEndianUnion, LittleEndianUnion classes to ctypes

### PR DESCRIPTION
**Pull request by ringof**
https://bugs.python.org/issue33178

Adds the BigEndianUnion, LittleEndianUnion classes documented here: https://docs.python.org/3/library/ctypes.html#structure-union-alignment-and-byte-order but are unimplemented in ctypes.

Co-authored-by: syeberman (Sye van Der Veen)
Source for this PR is derived from a patch submitted by syeberman for [bpo-19023](https://bugs.python.org/issue19023): 
https://bugs.python.org/issue19023
https://bugs.python.org/file31769/75843d82f6cf.diff 
for which these changes were not included in the final patch used to close that issue.

https://bugs.python.org/issue33178

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).



## CLA Missing

Our records indicate the following people have not signed the CLA:

@ringof

For legal reasons we need all the people listed to sign the CLA before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

File: Lib/ctypes/test/test_byteswap.py
Comment (FFY00) [inline Lib/ctypes/test/test_byteswap.py]: IMO this test would be better named something like `test_union_fields_unsupported_byte_order`.

File: Lib/ctypes/test/test_byteswap.py
Comment (FFY00) [inline Lib/ctypes/test/test_byteswap.py]: The field appended here will not be removed and so will be available in the next iteration, which I assume is not what you want. I recommend using the `+` operator instead to avoid modifying the original list.

I would do something like this.
```python
fields = [
    ("a", c_ubyte),
    ("b", c_byte),
    ("c", c_short),
    ("d", c_ushort),
    ("e", c_int),
    ("f", c_uint),
    ("g", c_long),
    ("h", c_ulong),
    ("i", c_longlong),
    ("k", c_ulonglong),
    ("l", c_float),
    ("m", c_double),
    ("n", c_char),
    ("b1", c_byte, 3),
    ("b2", c_byte, 3),
    ("b3", c_byte, 2),
    ("a", c_int * 3 * 3 * 3),
]

for typ in (c_wchar, c_void_p, POINTER(c_int)):
    with self.assertRaises(TypeError):
        class T(base):
            _fields_ = fields + [("x", typ)]
```

Comment (ringof) [reply Lib/ctypes/test/test_byteswap.py]: Thanks for the review!

https://github.com/python/cpython/blob/8d28e0a61edeb073406c03a102b1145542f7da63/Lib/ctypes/test/test_byteswap.py#L173-L206

test_struct_field_1 is nearly the same as the code you raised this comment on; will fix for the same issue.

**Review by FFY00 — APPROVED** [review]
Everything looks alright to me now :blush:

**Comment by FFY00** [comment]
The only improvement I could suggest at the moment is maybe using subtests so that we can easily distinguish what failed if the tests break. But I'll leave that to the person who will do the final review.

https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests

**Comment by FFY00** [comment]
@iritkatriel since you have been active lately, would you mind having a look at this? No worries if you are not comfortable enough to review it.

(let me know if you don't want me to ping you like this in the future)

**Comment by iritkatriel** [comment]
@FFY00 I don't mind the ping, but I also don't know much at all about ctypes.

**Comment by ringof** [comment]
belopolsky, amaury.forgeotdarc, meador.inge (B.P.O usernames) are listed as the active experts on ctypes in the Python Developer's Guide. I don't know which are current core developers.

**Comment by ringof** [comment]
@pablogsal any suggestions on a core developer to review this PR?

**Comment by shawwn** [comment]
@FFY00 @ringof @pablogsal What's the status of this PR? Projects such as [pyembc](https://github.com/waszil/pyembc) could benefit from `BigEndianUnion` and `LittleEndianUnion`:

https://github.com/waszil/pyembc/blob/52fd5c13317f822e8ef0a330c4bcbb7f2d7e0521/pyembc/_pyembc.py#L297-L312

```py
    # ctypes currently does not implement the BigEndianUnion and LittleEndianUnion despite its documentation
    # sais so. Therefore, we use simple Union for now. Details:
    # https://stackoverflow.com/questions/49524952/bigendianunion-is-not-part-of-pythons-ctypes
    # https://bugs.python.org/issue33178
    if endian == "little":
        _bases = {
            _PyembcTarget.STRUCT: ctypes.LittleEndianStructure,
            _PyembcTarget.UNION: ctypes.Union
        }
    elif endian == "big":
        _bases = {
            _PyembcTarget.STRUCT: ctypes.BigEndianStructure,
            _PyembcTarget.UNION: ctypes.Union
        }
    else:
        raise ValueError("Invalid endianness")
```

I discovered this PR via https://bugs.python.org/issue33178#msg397291:

> Author: David Goncalves (dpg) * | Date: 2021-07-12 09:37
> 
> PR 25480 passes checks and awaits core review. All comments thus far have been addressed.

So there appears to be no activity on this since July 2021. (In a few days it'll be January 2022.)

Is there any way I can help out? What would be the most productive thing I could do to help get this merged?

Thanks for your time!

**Comment by ringof** [comment]
@shawwn its awaiting core developer review - I wasn't able to find one with the experience and bandwidth to review this. I'm willing to do my part to respond to requests for changes.

**Comment by ringof** [comment]
The following list of names is from recently closed PRs or PRs with a glancing 

@iritkatriel @tiran @ericsnowcurrently @vstinner @pablogsal @markshannon @ambv @serhiy-storchaka 

I'm looking for a core review for this PR, but the Experts list does not suggest anybody who is recently active. Anybody have a suggestions who to reach out to do so?

File: Lib/ctypes/test/test_byteswap.py
Comment (gpshead) [inline Lib/ctypes/test/test_byteswap.py]: Consider writing these 6 lines as one:

```python
class T(BigEndianUnion if sys.byteorder == "little" else LittleEndianUnion):
```

File: Doc/library/ctypes.rst
Comment (gpshead) [inline Doc/library/ctypes.rst:2393]: these two need versionadded 3.11 markers in the docs.

Comment (ringof) [reply Doc/library/ctypes.rst:2393]: Agreed.

File: Lib/ctypes/test/test_byteswap.py
Comment (ringof) [reply Lib/ctypes/test/test_byteswap.py]: Agreed. 

Would you agree with applying the same approach to line 173: test_struct_fields_unsupported_byte_order for consistency?

Comment (gpshead) [reply Lib/ctypes/test/test_byteswap.py]: Sure. This is one of those "either way works" cases, not a big deal. The existing code likely predates the inline conditional syntax so it was never considered when written.

**Comment by ringof** [comment]
@gpshead All your comments are addressed. I'm not sure about the line length for what I did for the class definitions in the unsupported types tests - is there a general maximum and preferred line breaking style, I'll be glad to fix that.

**Review by gpshead — APPROVED** [review]

**Comment by ringof** [comment]
Thanks - glad it made it!

**Pull request merged**
